### PR TITLE
Introduce primitives for input/output dtype conversion in Lite Precision

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

The Precision plugin exposes API to convert model input and outputs to the appropriate precision dtype. 
Note that this is just for Lite for now, as precision integration in PL is still ongoing. 

Follow up to #14815

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @awaelchli @rohitgr7 @carmocca @akihironitta

### CODE REVIEW

File: src/lightning_lite/plugins/precision/deepspeed.py [hunk 1]
```diff
@@ -53,18 +56,23 @@ def __init__(self, precision: Union[str, int], amp_type: str, amp_level: Optiona
 
             amp_level = amp_level or "O2"
 
-        supported_precision = (PrecisionType.HALF, PrecisionType.FLOAT, PrecisionType.BFLOAT)
-        if precision not in supported_precision:
```
Comment (awaelchli) [inline src/lightning_lite/plugins/precision/deepspeed.py:57]: This was a bug, right? Comparing the precision (int) with the enum strings would never work here, i.e., the if-condition would be false.

File: src/lightning_lite/wrappers.py [hunk 2]
```diff
@@ -96,26 +96,12 @@ def module(self) -> nn.Module:
...
         """Casts all inputs to the right precision and handles autocast for operations in the module forward
         method."""
-        precision = self._precision_plugin.precision
-        precision_to_type = {
-            "bf16": torch.bfloat16,
-            16: torch.float16,
-            32: torch.float32,
-            64: torch.float64,
-        }
-        # TODO: let the precision plugin handle the conversion
```
Comment (awaelchli) [inline src/lightning_lite/wrappers.py:106]: This is the TODO we want to resolve in this PR.

File: src/pytorch_lightning/strategies/utils.py [hunk 3]
```diff
@@ -34,3 +37,13 @@ def _call_register_strategies(registry: _StrategyRegistry, base_module: str) ->
     for _, mod in getmembers(module, isclass):
         if issubclass(mod, Strategy) and _is_register_method_overridden(mod, Strategy, "register_strategies"):
             mod.register_strategies(registry)
+
+
+def _fp_to_half(tensor: torch.Tensor, precision: PrecisionType) -> torch.Tensor:
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/utils.py:42]: Moves back to where it came from.

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 1 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/deepspeed.py:57]: No. Our enums are magic :sparkles: and convert to str in the equality check: https://github.com/Lightning-AI/utilities/blob/86e9e48e453ae9690448a0d3cef320d3dc4bd777/src/lightning_utilities/core/enums.py#L23

File: src/lightning_lite/plugins/precision/deepspeed.py [hunk 4]
```diff
@@ -53,18 +56,23 @@ def __init__(self, precision: Union[str, int], amp_type: str, amp_level: Optiona
...
+        supported_precision = ("16", "32", "bf16")
+        if str(precision) not in supported_precision:
             raise ValueError(
                 f"`precision={precision!r})` is not supported in DeepSpeed."
-                f" `precision` must be one of: {(x.value for x in supported_precision)}."
+                f" `precision` must be one of: {', '.join(supported_precision)}."
             )
 
         super().__init__()
         self.precision = precision
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/deepspeed.py:67]: If you stop using the enums to check. Then IMO this should be saved as a string to avoid potential issues with checking `str == int`

`self.precision = str(self.precision)`

And in the dictionaries: `{"bf16": torch.bfloat16, "16": torch.float16, "32": torch.float32}`

**Review by carmocca — APPROVED** [review]
Merged #14823

File: src/lightning_lite/plugins/precision/precision.py [hunk 5]
```diff
@@ -34,6 +36,20 @@ def forward_context(self) -> Generator[None, None, None]:
         """A contextmanager for managing model forward/training_step/evaluation_step/predict_step."""
         yield
 
+    def convert_input(self, data: Tensor) -> Tensor:
+        """Convert model inputs (forward) to the floating point precision type of this plugin.
+
+        This is a no-op for tensors that are not of floating-point type or already have the desired type.
+        """
+        return _convert_fp_tensor(data, torch.float32)
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/precision.py:44]: Couldn't we do this in the base

```suggestion
        precision_to_type = {"bf16": torch.bfloat16, 16: torch.float16, 32: torch.float32}
        dst_type = precision_to_type[self.precision]
        return _convert_fp_tensor(data, dst_type)
```

so that none of the subclasses need to override it?

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/deepspeed.py:67]: Yes, I had the impression we decided to move away from the enums a while ago, although I don't remember all the arguments. 

I agree it would be good to only have the string type. I need to check where the precision attribute is used, connector etc.

File: src/lightning_lite/plugins/precision/precision.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/precision.py:44]: Ideally, I want to mark theses interfaces abstract to force all of them to implement it. But I can't as this would bring more breaking changes on the PL side. 

Since some plugins can only manage a subset of precision types, I think it would be wrong if the super implementation could handle more than allowed.

File: src/pytorch_lightning/strategies/utils.py (hunk 3 above)
Comment (awaelchli) [reply src/pytorch_lightning/strategies/utils.py:43]: Todo @awaelchli 
Remove after Carlos changes

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/deepspeed.py:57]: That explains it .Thanks

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 4 above)
Comment (justusschock) [reply src/lightning_lite/plugins/precision/deepspeed.py:67]: See #12069 which is related

File: src/lightning_lite/plugins/precision/precision.py (hunk 5 above)
Comment (justusschock) [reply src/lightning_lite/plugins/precision/precision.py:44]: I prefer having that as abstract interface.

File: src/lightning_lite/plugins/precision/precision.py [hunk 6]
```diff
@@ -34,6 +36,20 @@ def forward_context(self) -> Generator[None, None, None]:
...
+
+    def convert_output(self, data: Tensor) -> Tensor:
```
Comment (justusschock) [inline src/lightning_lite/plugins/precision/precision.py:46]: what if the input was not of default type? Will we still convert the output to default type or should that rather be converted to the inputs type?

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 4 above)
Comment (justusschock) [reply src/lightning_lite/plugins/precision/deepspeed.py:67]: @carmocca @awaelchli while on it and while adding a new package, should we also integrate https://github.com/Lightning-AI/lightning/issues/9956#issuecomment-1207246337 to avoid the deprecation cycle here?

File: src/lightning_lite/plugins/precision/precision.py (hunk 6 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/precision.py:46]: Good point Justus. Maybe we should not convert the output at all, or do the conversion back to input as you said? For example, using double precision, we would downgrade to lower precision (32) but maybe user needs the double for further loss calculations. 

In both these cases, we wouldn't need a convert_output() method and would handle it directly in the lite forward wrapper.

Comment (awaelchli) [reply src/lightning_lite/plugins/precision/precision.py:46]: In any case, right now the implementation is in parity to what was before. But no saying the previous way was more correct.

File: src/lightning_lite/plugins/precision/precision.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/precision.py:44]: Some options:

1. Current
2. Add abstract now: Small breaking change for custom plugins (who makes them?). Plus, when #14798 merges, we need to provide the implementations for PL plugins
3. Add `TODO(lite): make abstract` to delay it

File: src/lightning_lite/connector.py [hunk 7]
```diff
@@ -58,6 +59,7 @@
 
 _PLUGIN = Union[Precision, ClusterEnvironment, CheckpointIO]
 _PLUGIN_INPUT = Union[_PLUGIN, str]
+_PRECISION_INPUT = Literal[16, 32, 64, "bf16"]
```
Comment (carmocca) [inline src/lightning_lite/connector.py:62]: I would also include `"16", "32", "64"`

File: src/lightning_lite/plugins/precision/deepspeed.py [hunk 8]
```diff
@@ -43,7 +46,7 @@ class DeepSpeedPrecision(Precision):
             If unsupported ``precision`` is provided.
     """
 
-    def __init__(self, precision: Union[str, int], amp_type: str, amp_level: Optional[str] = None) -> None:
+    def __init__(self, precision: Literal[16, 32, "bf16"], amp_type: str, amp_level: Optional[str] = None) -> None:
```
Comment (carmocca) [inline src/lightning_lite/plugins/precision/deepspeed.py:49]: Same suggestion

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 4 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/deepspeed.py:67]: > I had the impression we decided to move away from the enums a while ago, although I don't remember all the arguments.

The argument is that they are not easy to extend.

Let's continue the discussion about planning for #9956 offline.

> I agree it would be good to only have the string type. I need to check where the precision attribute is used, connector etc.

From this comment I assume you still need to set it as a str

File: src/lightning_lite/plugins/precision/precision.py (hunk 5 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/precision.py:44]: If the PL classes will not define it. It's boilerplate-y to mark it as abstract

File: src/lightning_lite/plugins/precision/precision.py (hunk 6 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/precision.py:46]: The safest thing IMO would be to convert to the original data precision. Otherwise not do this at all.

Can you create an issue about this for tracking?

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 8 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/deepspeed.py:49]: Regarding this and the other comments, I spent like an hour on this and it's spreading like a cancer. I have to break it out because the mypy is just forcing me to do it everywhere. 
#14857

File: src/lightning_lite/connector.py (hunk 7 above)
Comment (awaelchli) [reply src/lightning_lite/connector.py:62]: Please check my comment below. If I include string alternatives here, now I need to update the comparisons in the connector, which then leads to more typing updates, which then leads to etc. etc.

File: src/lightning_lite/plugins/precision/precision.py (hunk 6 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/precision.py]: Created issue #14859

Removed the convert_output method. We can introduce it later should we have a decision. Right now, I'll just leave the output as it was handled previously.

File: src/lightning_lite/plugins/precision/deepspeed.py (hunk 8 above)
Comment (carmocca) [reply src/lightning_lite/plugins/precision/deepspeed.py:49]: If it's becoming a larger unnecessary refactor, you could also revert back to Union[str, int] and still use the `PrecisionType` magic enums for equalities

Comment (awaelchli) [reply src/lightning_lite/plugins/precision/deepspeed.py:49]: yes, reverted. using old logic now. can be re-done in https://github.com/Lightning-AI/lightning/pull/14857

**Review by carmocca — APPROVED** [review]
Re-approving after the latest changes

File: tests/tests_lite/plugins/precision/test_native_amp_integration.py [hunk 9]
```diff
@@ -67,6 +67,6 @@ def after_backward(self, model):
     ],
 )
 def test_native_mixed_precision(accelerator, precision, expected_dtype):
-    lite = NativeMixedPrecisionBoringLite(accelerator=accelerator, precision=16)
+    lite = NativeMixedPrecisionBoringLite(accelerator=accelerator, precision=precision)
```
Comment (awaelchli) [inline tests/tests_lite/plugins/precision/test_native_amp_integration.py:70]: Including a small fix here. This was failing locally on the rtx card where bf16_cuda is available (not here in CI). And the test was obviously meant to take the parametrized argument as input.

**Review by otaj — APPROVED** [review]
